### PR TITLE
add network interface tune support

### PR DESCRIPTION
**Pull request by larryliuqing**
Signed-off-by: Larry Liu <larryliuqing@163.com>

**What this PR does / why we need it**:
This  PR add network interface tune support based on Libvirt https://libvirt.org/formatnetwork.html#quality-of-service

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

``` 
add network interface tune support

```

### CODE REVIEW

**Comment by rmohr** [comment]
Similar to https://github.com/kubevirt/kubevirt/pull/7605#issuecomment-1104847699 I think we would need a discussion upfront before we could add something like this.

**Comment by larryliuqing** [comment]
Based on the https://github.com/kubevirt/kubevirt/pull/7605#issuecomment-1104853173, the suggestion solution uses sidecar hook to implement the libvirt parameters hack, but the network interface tune needs the NET_ADMIN capability for virt-lancher-xxx pod, then I want to know if there is some way to add the capability ?

File: pkg/virt-controller/services/template.go [hunk 1]
```diff
@@ -1825,6 +1826,12 @@ func getRequiredCapabilities(vmi *v1.VirtualMachineInstance) []k8sv1.Capability
 	// https://github.com/libvirt/libvirt/commit/a9c500d2b50c5c041a1bb6ae9724402cf1cec8fe
 	capabilities = append(capabilities, CAP_SYS_PTRACE)
 
+	for _, nic := range vmi.Spec.Domain.Devices.Interfaces {
+		if nic.Outbound != nil || nic.Inbound != nil {
+			capabilities = append(capabilities, CAP_NET_ADMIN)
```
Comment (xpivarc) [inline pkg/virt-controller/services/template.go:1831]: We don't want to introduce new capabilities. We worked hard to remove them.

**Review by xpivarc — COMMENTED** [review]
/hold
I think this would be more harmful than beneficial. We don't want to introduce new capabilities.

**Comment by EdDev** [comment]
> Based on the [#7605 (comment)](https://github.com/kubevirt/kubevirt/pull/7605#issuecomment-1104853173), the suggestion solution uses sidecar hook to implement the libvirt parameters hack, but the network interface tune needs the NET_ADMIN capability for virt-lancher-xxx pod, then I want to know if there is some way to add the capability ?

@larryliuqing , kubevirt is focused on integrating as much as possible with kubernetes environment, utilizing it services over doing the same in the (old/classic) way. It originates from the need to integrate with the existing networking solutions k8s exposes to pods in general and with our security needs to keep the pod in which the VM runs with the lowest capability set as possible (and therefore secured).

In general, I think you should look for k8s native solutions over the libvirt/virt ones. As an example, the [bandwidth CNI](https://www.cni.dev/plugins/current/meta/bandwidth/) is the closest to what you are looking for. You should prioritize trying it out over hacking your way with sidecars.

There have been other implementation which have been inspired from libvirt and applied as CNI plugin options (e.g. [macspoof](https://kubevirt.io/user-guide/virtual_machines/interfaces_and_networks/#mac-spoof-check)). I think this is no different.
Either use the existing meta CNI ref before or extend it to add the part you are missing.
By that, you will not only contribute to kubevirt, but to the container ecosystem in general.

You should also note that some network providers (e.g. OVN, Calico) may already have such solutions in they arsenal, something that may or may not fit your needs. But at least worth exploring.

If you would like to track this issue further and share your insights, I suggest to open an issue.

**Comment by larryliuqing** [comment]
@EdDev : Thanks very much that give me so many suggestion and more cloud native mind to build the missing functions, I will try these suggestions, and feed back to community when I have better idea/solution.

Thanks every one again!

**Pull request closed without merging**